src/Strategy/Maven/DepTree.hs

```diff
@@ -0,0 +1,180 @@
+analyze dir = do
+  -- We generate files with a random prefix so we don't pick up files from previous runs.
+  -- If we did pick them up, they're likely not enough of a conflict to worry about, but
+  -- it's better than not doing it.
+  randIdent <- abs <$> sendIO randomIO
+  _ <- context "Running maven 'dependency:tree' plugin" $ execThrow dir $ deptreeCmd randIdent
+  graphFiles <- context "Locating maven output files" $ findDepTreeOutputs dir randIdent
+  graphs <- context "Parsing output files" $ traverse (readContentsParser parseDotGraph) graphFiles
+  pure $ buildGraph graphs
+
+findDepTreeOutputs :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> Int -> m [Path Abs File]
+findDepTreeOutputs dir ident = execState @[Path Abs File] [] $
+  flip walk dir $ \_ _ files -> do
+    case findFileNamed (show ident <> outputFileName) files of
+      Nothing -> pure WalkContinue
+      Just file -> do
+        modify (file :)
+        pure WalkContinue
+
+buildGraph :: [DotGraph] -> Graphing Dependency
+buildGraph = gmap toDependency . foldr ((<>) . stripRoot . toGraph) mempty
+
+toDependency :: PackageId -> Dependency
+toDependency PackageId{..} =
+  Dependency
+    { dependencyType = MavenType
+    , dependencyName = artifactName
+    , dependencyVersion = Just $ CEq artifactVersion
+    , dependencyLocations = []
+    , dependencyEnvironments = maybe [EnvProduction] ((: []) . toBuildTag) buildTag
+    , dependencyTags = mempty
+    }
+
+toGraph :: DotGraph -> Graphing PackageId
+toGraph DotGraph{..} = run . evalGrapher $ do
+  direct rootNode
+  for_ edgeList $ uncurry edge
+
+toBuildTag :: Text -> DepEnvironment
+toBuildTag = \case
+  "compile" -> EnvProduction
+  "test" -> EnvTesting
+  other -> EnvOther other
+
+deptreeCmd :: Int -> Command
+deptreeCmd ident =
+  Command
+    { cmdName = "mvn"
+    , cmdArgs =
+        [ "dependency:tree" -- run the command
+        , "-DoutputType=dot" -- export DOT format
+        , "-DoutputFile=" <> toText (show ident) <> outputFileName -- dump files to these names in the relevant directories.
+        ]
+    , cmdAllowErr = Never
+    }
+
+data PackageId = PackageId
+  { groupName :: Text
+  , artifactName :: Text
+  , artifactType :: Text
+  , artifactVersion :: Text
+  , buildTag :: Maybe Text
+  }
+  deriving (Eq, Ord, Show)
+
+type Parser = Parsec Void Text
+
+data DotGraph = DotGraph
+  { rootNode :: PackageId
+  , edgeList :: [(PackageId, PackageId)]
+  }
+  deriving (Eq, Ord, Show)
+
+parseDotGraph :: Parser DotGraph
+parseDotGraph = do
+  root <- symbol "digraph" *> parseNode
+  edgeLists <- enclosed "{" "}" (many $ try parseGraphEntry)
+  pure . DotGraph root $ concatMap sequenceEdges edgeLists
+
+-- | Produce tuples of every adjacent pair in the list.
+-- 1-item lists contain no edges, and are discarded.
+--
+-- >>> sequenceEdges [1, 2, 3, 4]
+-- [(1, 2), (2, 3), (3, 4)]
+-- >>> sequenceEdges [1]
+-- []
+sequenceEdges :: [a] -> [(a, a)]
```

Comment:
drive-by clever snippet: we can implement this with `zip`

```hs
sequenceEdges :: [a] -> [(a, a)]
sequenceEdges xs = zip xs (drop 1 xs)
```

..because `drop 1` will align the next element in the list with the current element for the zip:
```
xs        = [1, 2, 3, 4]
drop 1 xs = [2, 3, 4]
```

Reply:
👏 👏 👏